File: src/main/java/io/vproxy/vfx/util/algebradata/XYZData.java
package io.vproxy.vfx.util.algebradata;

public class XYZData implements AlgebraData<XYZData> {
    public final double x;
    public final double y;
    public final double z;

    public XYZData(double x, double y, double z) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    @Override
    public XYZData plus(XYZData other) {
        return new XYZData(x + other.x, y + other.y, z + other.z);
    }

    @Override
    public XYZData minus(XYZData other) {
        return new XYZData(x - other.x, y - other.y, z - other.z);
    }

    @Override
    public XYZData multiply(double v) {
        return new XYZData(x * v, y * v, z * v);
    }

    @Override
    public XYZData dividedBy(double v) {
        return new XYZData(x / v, y / v, z / v);
    }

    @Override
    public String toString() {
        return "XYZData(" + x + ", " + y + ", " + z + ")";
    }
}

File: src-ts/src/util/algebradata/XYZData.ts
import { AlgebraData } from './AlgebraData.js';

export class XYZData implements AlgebraData<XYZData> {
  constructor(public readonly x: number, public readonly y: number, public readonly z: number) {}

  plus(o: XYZData): XYZData { return new XYZData(this.x + o.x, this.y + o.y, this.z + o.z); }
  minus(o: XYZData): XYZData { return new XYZData(this.x - o.x, this.y - o.y, this.z - o.z); }
  multiply(v: number): XYZData { return new XYZData(this.x * v, this.y * v, this.z * v); }
  dividedBy(v: number): XYZData { return new XYZData(this.x / v, this.y / v, this.z / v); }

  toString(): string { return `XYZData(${this.x}, ${this.y}, ${this.z})`; }
}
